src/instance.rs

```diff
@@ -954,6 +961,35 @@ impl<'a, 'py, T> Borrowed<'a, 'py, T> {
     {
         FromPyObject::extract(self.to_any())
     }
+
+    /// Cast this to a concrete Python type or pyclass.
+    ///
+    /// This performs a runtime type check using the equivalent of Python's
+    /// `isinstance(self, U)`.
+    #[inline]
+    pub fn cast<U>(self) -> Result<Borrowed<'a, 'py, U>, DowncastError<'a, 'py>>
+    where
+        U: PyTypeCheck,
+    {
+        if U::type_check(&self.to_any()) {
+            // Safety: type_check is responsible for ensuring that the type is correct
+            Ok(unsafe { self.cast_unchecked() })
+        } else {
+            Err(DowncastError::new_from_type(
+                self.to_any(),
+                U::classinfo_object(self.py()),
+            ))
+        }
+    }
+
```

Comment:
There is probably also space for a `.cast_exact` (even if we don't use it internally yet).

Reply:
Good idea 👍